### PR TITLE
Quick script to look at whois requests via the Finger analyzer

### PR DESCRIPTION
**Pull request by grigorescu**
It seemed like a waste that the Finger analyzer is part of Zeek, but nothing uses it. Rather than focus on the lesser-used finger protocol, I thought it'd be more interesting to log what is being queried via "whois" on your network. This includes domain names and IP addresses.

I can see a couple of operational use-cases for this. We've noticed attackers running a combo of "whatismyip / whois IP" once they get a shell on a compromised system. Additionally, some of our more suspicious users run whois for a domain name they're being asked to click on. In any case, it's network traffic that's going across the wire, and it seemed fitting to log it.

I chose not to log the response. The protocol will send a number of newline-delimited replies (often a few dozen), which are not formatted consistently. The analyzer currently generates a new response event for each of these lines. I suppose someone could throw these into the file analysis framework, but I'm not convinced they're operationally very useful today. At most some kind of "passive whois" datastore.

### CODE REVIEW

File: scripts/base/protocols/whois/main.zeek [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+module Whois;
+
+export {
+	## The Whois protocol logging stream identifier.
+	redef enum Log::ID += { LOG };
+
+	## A default logging policy hook for the stream.
+	global log_policy: Log::PolicyHook;
+
+	## The record type which contains the column fields of the DHCP log.
```
Comment (jsiwek) [inline scripts/base/protocols/whois/main.zeek:13]:
```suggestion
	## The record type which contains the column fields of the WHOIS log.
```

**Review by jsiwek — APPROVED** [review]
Thanks, looks good / sounds interesting to me at least.  Seems like it could also be easily maintained an external package, so wanted to check if there's any latest opinions/clarifications on what criteria makes something like this better to maintain in Zeek versus a package ?

**Comment by rsmmr** [comment]
For new optional functionality, I'd say packages are preferable these days. If this were a general analyzer script for `finger` (which is indeed missing), I could see merging it in. Given that it's more specific than that, a package seems the better fit to me.

 @grigorescu would the main reason for including into the distribution be  the fact that finger isn't having any scripts right now?

**Comment by grigorescu** [comment]
> @grigorescu would the main reason for including into the distribution be the fact that finger isn't having any scripts right now?

Yes, what got me started on this was the recent work on code coverage stats on the Zeek code, and the realization that some analyzers had 0 coverage. While I do have a script that logs finger usage, I ran that script from 2012-2014, at a rare institution that still relied heavily on finger, and found it of very limited use operationally. I guess I could submit a PR for that, but I really don't think it's useful today. Whois, on the other hand, has a surprising amount of operational utility, as I mentioned above.

My view is that if the Finger analyzer is in the core, this script belongs in base. I know it's not the same protocol, but the Finger analyzer is really a Whois analyzer with support for the Finger extensions to the protocol. I see this as correcting a gap in Zeek logging. I could see an argument for moving the Finger analyzer to a plugin, but given that the protocol hasn't changed in almost 30 years, I don't think the maintenance risk/overhead is that high.

**Comment by rsmmr** [comment]
>  Finger analyzer is really a Whois analyzer with support for the Finger extensions to the protocol. Finger feels like an anachronism today, and the analyzer does not support the full Finger RFC.

That may be the key point. Should we rename the whole analyzer "WhoIs" then? Then it would be more clear why this belongs into base.

**Comment by grigorescu** [comment]
> Should we rename the whole analyzer "WhoIs" then? Then it would be more clear why this belongs into base.

Yeah, I like this. There is some cruft in the analyzer itself, as well as stuff like the finger content matching in signatures.

I see two options for how to proceed:

1. We could write a whois analyzer, with only the code necessary to analyze that protocol (essentially a stripped-down version of the finger analyzer). Then we remove the finger analyzer, as well as the finger content matching in the signature framework. We could remove this with our standard deprecation cycle, or not. If anyone wants the finger functionality in the future, they could dig the code out of git.
 
2. We rename the current finger analyzer to a whois analyzer, leaving the finger-specific code paths in place in case anyone wants to use those in the future. Maybe we add an event with a better name (e.g. `whois_request`).

**Comment by jsiwek** [comment]
> 1. We could write a whois analyzer, with only the code necessary to analyze that protocol (essentially a stripped-down version of the finger analyzer). Then we remove the finger analyzer, as well as the finger content matching in the signature framework. We could remove this with our standard deprecation cycle, or not. If anyone wants the finger functionality in the future, they could dig the code out of git.

This is a path that makes sense to me provided the current Finger analyzer isn't really complete in the first place and there's not any earnest intention to flesh it out and include tests cases to cover all of it.  Maybe can also do a quick sanity check on mailing-list/Slack to see if anyone is vocal about using the current analyzer and would be annoyed by the change.

**Comment by grigorescu** [comment]
> Maybe can also do a quick sanity check on mailing-list/Slack to see if anyone is vocal about using the current analyzer and would be annoyed by the change.

Done. I'm going to leave this open until early next week in case anyone wants to chime in here, then plan on closing it.

**Comment by jsiwek** [comment]
Closing since it seems like the favorable path would be to convert the finger analyzer to specific whois analyzer (justifies adding `whois.log` as part of base scripts).

**Comment by grigorescu** [comment]
I did get some feedback that someone does use the Finger analyzer, but I think I can make the whois change separately from that.

**Pull request closed without merging**